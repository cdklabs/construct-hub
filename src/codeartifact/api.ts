/**
 * The CodeArtifact repository API exposed to extensions.
 */
export interface IRepository {
  /**
   * Adds an external connection to this repository.
   *
   * @param id the id of the external connection (i.e: `public:npmjs`).
   */
  addExternalConnection(id: string): void;
}
