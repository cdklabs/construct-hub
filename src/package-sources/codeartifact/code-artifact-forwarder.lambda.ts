import { metricScope, Unit } from 'aws-embedded-metrics';
import type { Context, EventBridgeEvent } from 'aws-lambda';

import { DenyListClient } from '../../backend/deny-list/client.lambda-shared';
import { LicenseListClient } from '../../backend/license-list/client.lambda-shared';
import * as aws from '../../backend/shared/aws.lambda-shared';
import { requireEnv } from '../../backend/shared/env.lambda-shared';
import { integrity } from '../../backend/shared/integrity.lambda-shared';
import { extractObjects } from '../../backend/shared/tarball.lambda-shared';
import {
  METRICS_NAMESPACE,
  MetricName,
  DOMAIN_OWNER_DIMENSION,
  DOMAIN_NAME_DIMENSION,
  REPOSITORY_NAME_DIMENSION,
} from './constants.lambda-shared';

const DETAIL_TYPE = 'CodeArtifact Package Version State Change' as const;

export const handler = metricScope(
  (metrics) =>
    async (
      event: EventBridgeEvent<typeof DETAIL_TYPE, CodeArtifactDetail>,
      context: Context
    ) => {
      console.log(`Event: ${JSON.stringify(event, null, 2)}`);

      const stagingBucket = requireEnv('BUCKET_NAME');
      const queueUrl = requireEnv('QUEUE_URL');

      metrics.setNamespace(METRICS_NAMESPACE);
      metrics.setDimensions({
        [DOMAIN_OWNER_DIMENSION]: event.detail.domainOwner,
        [DOMAIN_NAME_DIMENSION]: event.detail.domainName,
        [REPOSITORY_NAME_DIMENSION]: event.detail.repositoryName,
      });

      const packageName = event.detail.packageNamespace
        ? `@${event.detail.packageNamespace}/${event.detail.packageName}`
        : event.detail.packageName;

      if (event.detail.operationType === 'Deleted') {
        console.log(
          `Operation type is "Deleted" for ${packageName}@${event.detail.packageVersion}`
        );
        metrics.putMetric(MetricName.DELETED_COUNT, 1, Unit.Count);
        return;
      }

      const denyList = await DenyListClient.newClient();
      const denied = denyList.lookup(packageName, event.detail.packageVersion);
      if (denied) {
        console.log(
          `Package ${packageName}@${
            event.detail.packageVersion
          } denied: ${JSON.stringify(denied, null, 2)}`
        );
        metrics.putMetric(MetricName.DENY_LISTED_COUNT, 1, Unit.Count);
        return;
      }

      const { asset, packageVersionRevision } = await aws
        .codeArtifact()
        .getPackageVersionAsset({
          asset: 'package.tgz', // Always named this way for npm packages!
          domainOwner: event.detail.domainOwner,
          domain: event.detail.domainName,
          repository: event.detail.repositoryName,
          format: event.detail.packageFormat,
          namespace: event.detail.packageNamespace,
          package: event.detail.packageName,
          packageVersion: event.detail.packageVersion,
        })
        .promise();
      const tarball = Buffer.from(asset! as any);

      const { assemblyJson, packageJson } = await extractObjects(tarball, {
        assemblyJson: { path: 'package/.jsii' },
        packageJson: { path: 'package/package.json', required: true },
      });
      metrics.putMetric(
        MetricName.NOT_JSII_ENABLED_COUNT,
        assemblyJson ? 0 : 1,
        Unit.Count
      );
      if (assemblyJson == null) {
        console.log(
          `Package "${packageName}@${event.detail.packageVersion}" does not contain a .jsii assembly`
        );
        return;
      }

      const metadata = JSON.parse(packageJson.toString('utf-8'));
      const licenseList = await LicenseListClient.newClient();
      const eligibleLicense = licenseList.lookup(
        metadata.license ?? 'UNLICENSED'
      );
      metrics.putMetric(
        MetricName.INELIGIBLE_LICENSE,
        eligibleLicense ? 0 : 1,
        Unit.Count
      );
      if (!eligibleLicense) {
        console.log(
          `Package "${packageName}@${
            event.detail.packageVersion
          }" does not use allow-listed license: ${
            metadata.license ?? 'UNLICENSED'
          }`
        );
        return;
      }

      const stagingKey = `${packageName}/${event.detail.packageVersion}/${packageVersionRevision}/package.tgz`;
      await aws
        .s3()
        .putObject({
          Bucket: stagingBucket,
          Key: stagingKey,
          Body: asset,
          ContentType: 'application/octet-stream',
          Metadata: {
            'Lambda-Log-Group': context.logGroupName,
            'Lambda-Log-Stream': context.logStreamName,
            'Lambda-Run-Id': context.awsRequestId,
          },
        })
        .promise();

      const message = integrity(
        {
          tarballUri: `s3://${stagingBucket}/${stagingKey}`,
          metadata: { resources: event.resources.join(', ') },
          time: event.time,
        },
        tarball
      );
      return aws
        .sqs()
        .sendMessage({
          MessageAttributes: {
            AWS_REQUEST_ID: {
              DataType: 'String',
              StringValue: context.awsRequestId,
            },
            LOG_GROUP_NAME: {
              DataType: 'String',
              StringValue: context.logGroupName,
            },
            LOG_STREAM_NAME: {
              DataType: 'String',
              StringValue: context.logStreamName,
            },
          },
          MessageBody: JSON.stringify(message),
          QueueUrl: queueUrl,
        })
        .promise();
    }
);

/**
 * @see https://docs.aws.amazon.com/codeartifact/latest/ug/service-event-format-example.html
 */
interface CodeArtifactDetail {
  /**
   * The domain that contains the repository that contains the package.
   */
  readonly domainName: string;

  /**
   * The AWS account ID of the owner of the domain.
   */
  readonly domainOwner: string;

  /**
   * The repository that contains the package.
   */
  readonly repositoryName: string;

  /**
   * The format of the package that triggered the event.
   */
  readonly packageFormat: string;

  /**
   * The namespace of the package that triggered the event.
   */
  readonly packageNamespace?: string;

  /**
   * The name of the package that triggered the event.
   */
  readonly packageName: string;

  /**
   * The version of the package that triggered the event.
   */
  readonly packageVersion: string;

  /**
   * The state of the package version when the event was triggered.
   */
  readonly packageVersionState:
    | 'Unfinished'
    | 'Published'
    | 'Unlisted'
    | 'Archived'
    | 'Disposed';

  /**
   * A value that uniquely identifies the state of the assets and metadata of the package version when the event was
   * triggered. If the package version is modified (for example, by adding another JAR file to a Maven package), the
   * packageVersionRevision changes.
   */
  readonly packageVersionRevision: string;

  readonly changes: {
    /**
     * The number of assets added to a package that triggered an event. Examples of an asset are a Maven JAR file or a
     * Python wheel.
     */
    readonly assetsAdded: number;

    /**
     * The number of assets removed from a package that triggered an event.
     */
    readonly assetsRemoved: number;

    /**
     * The number of assets modified in the package that triggered the event.
     */
    readonly metadataUpdated: boolean;

    /**
     * A boolean value that is set to true if the event includes modified package-level metadata. For example, an event
     * might modify a Maven pom.xml file.
     */
    readonly assetsUpdated: number;

    /**
     * A boolean value that is set to true if the event's packageVersionStatus is modified(for example, if
     * packageVersionStatus changes from Unfinished to Published).
     */
    readonly statusChanged: boolean;
  };

  /**
   * Describes the high-level type of the package version change.
   */
  readonly operationType: 'Created' | 'Updated' | 'Deleted';

  /**
   * An integer that specifies an event number for a package. Each event on a package increments the sequenceNumber so
   * events can be arranged sequentially. An event can increment the sequenceNumber by any integer number.
   */
  readonly sequenceNumber: number;

  /**
   * An ID used to differentiate duplicate EventBridge events. In rare cases, EventBridge might trigger the same rule
   * more than once for a single event or scheduled time. Or, it might invoke the same target more than once for a given
   * triggered rule.
   */
  readonly eventDeduplicationId: string;
}
