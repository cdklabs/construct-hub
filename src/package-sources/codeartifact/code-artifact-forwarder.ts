// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

export interface CodeArtifactForwarderProps extends lambda.FunctionOptions {
}

export class CodeArtifactForwarder extends lambda.Function {
  constructor(scope: Construct, id: string, props?: CodeArtifactForwarderProps) {
    super(scope, id, {
      description: 'package-sources/codeartifact/code-artifact-forwarder.lambda.ts',
      ...props,
      architecture: lambda.Architecture.ARM_64,
      runtime: lambda.Runtime.NODEJS_16_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '/code-artifact-forwarder.lambda.bundle')),
    });
  }
}