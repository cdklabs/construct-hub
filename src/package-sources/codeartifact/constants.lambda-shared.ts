export const METRICS_NAMESPACE = 'ConstructHub/PackageSource/CodeArtifact';

export const DOMAIN_OWNER_DIMENSION = 'DomainOwner';
export const DOMAIN_NAME_DIMENSION = 'DomainName';
export const REPOSITORY_NAME_DIMENSION = 'RepositoryName';

export const enum MetricName {
  DELETED_COUNT = 'DeletedCount',
  DENY_LISTED_COUNT = 'DenyListedCount',
  INELIGIBLE_LICENSE = 'IneligibleLicenseCount',
  NOT_JSII_ENABLED_COUNT = 'NotJsiiEnabledCount',
}
