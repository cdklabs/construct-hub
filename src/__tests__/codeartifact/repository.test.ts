import { MatchStyle } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { InterfaceVpcEndpointAwsService, Vpc } from '@aws-cdk/aws-ec2';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { App, CfnOutput, Stack } from '@aws-cdk/core';
import { Repository } from '../../codeartifact/repository';

test('basic usage', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new Repository(stack, 'Repo');

  // THEN
  expect(stack).toMatchTemplate({
    Resources: {
      RepoDomainC79FB030: {
        Type: 'AWS::CodeArtifact::Domain',
        Properties: {
          DomainName: 'c8d064061d1c8680a574cd5a9f9c9c69b475d41907',
        },
      },
      Repo: {
        Type: 'AWS::CodeArtifact::Repository',
        Properties: {
          DomainName: {
            'Fn::GetAtt': [
              'RepoDomainC79FB030',
              'Name',
            ],
          },
          RepositoryName: 'c8d064061d1c8680a574cd5a9f9c9c69b475d41907',
        },
      },
    },
  }, MatchStyle.SUPERSET);
});

test('using upstreams', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new Repository(stack, 'Repo', { domainExists: true, domainName: 'test-domain', upstreams: ['upstream-1', 'upstream-2'] });

  // THEN
  expect(stack).toMatchTemplate({
    Resources: {
      Repo: {
        Type: 'AWS::CodeArtifact::Repository',
        Properties: {
          DomainName: 'test-domain',
          RepositoryName: 'c8d064061d1c8680a574cd5a9f9c9c69b475d41907',
          Upstreams: ['upstream-1', 'upstream-2'],
        },
      },
    },
  }, MatchStyle.SUPERSET);
});

test('external connection', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new Repository(stack, 'Repo').addExternalConnection('public:npmjs');

  // THEN
  expect(stack).toMatchTemplate({
    Resources: {
      RepoDomainC79FB030: {
        Type: 'AWS::CodeArtifact::Domain',
        Properties: {
          DomainName: 'c8d064061d1c8680a574cd5a9f9c9c69b475d41907',
        },
      },
      Repo: {
        Type: 'AWS::CodeArtifact::Repository',
        Properties: {
          DomainName: {
            'Fn::GetAtt': [
              'RepoDomainC79FB030',
              'Name',
            ],
          },
          RepositoryName: 'c8d064061d1c8680a574cd5a9f9c9c69b475d41907',
          ExternalConnections: ['public:npmjs'],
        },
      },
    },
  }, MatchStyle.SUPERSET);
});

test('custom domain name', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new Repository(stack, 'Repo', { domainName: 'custom-domain' });

  // THEN
  expect(stack).toMatchTemplate({
    Resources: {
      RepoDomainC79FB030: {
        Type: 'AWS::CodeArtifact::Domain',
        Properties: {
          DomainName: 'custom-domain',
        },
      },
      Repo: {
        Type: 'AWS::CodeArtifact::Repository',
        Properties: {
          DomainName: {
            'Fn::GetAtt': [
              'RepoDomainC79FB030',
              'Name',
            ],
          },
          RepositoryName: 'c8d064061d1c8680a574cd5a9f9c9c69b475d41907',
        },
      },
    },
  }, MatchStyle.SUPERSET);
});

test('custom repository name', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new Repository(stack, 'Repo', { repositoryName: 'custom-repo' });

  // THEN
  expect(stack).toMatchTemplate({
    Resources: {
      RepoDomainC79FB030: {
        Type: 'AWS::CodeArtifact::Domain',
        Properties: {
          DomainName: 'c8d064061d1c8680a574cd5a9f9c9c69b475d41907',
        },
      },
      Repo: {
        Type: 'AWS::CodeArtifact::Repository',
        Properties: {
          DomainName: {
            'Fn::GetAtt': [
              'RepoDomainC79FB030',
              'Name',
            ],
          },
          RepositoryName: 'custom-repo',
        },
      },
    },
  }, MatchStyle.SUPERSET);
});

test('custom domain & repository name', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new Repository(stack, 'Repo', { domainName: 'custom-domain', repositoryName: 'custom-repo' });

  // THEN
  expect(stack).toMatchTemplate({
    Resources: {
      RepoDomainC79FB030: {
        Type: 'AWS::CodeArtifact::Domain',
        Properties: {
          DomainName: 'custom-domain',
        },
      },
      Repo: {
        Type: 'AWS::CodeArtifact::Repository',
        Properties: {
          DomainName: {
            'Fn::GetAtt': [
              'RepoDomainC79FB030',
              'Name',
            ],
          },
          RepositoryName: 'custom-repo',
        },
      },
    },
  }, MatchStyle.SUPERSET);
});

test('npm repository endpoint', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  // WHEN
  const repo = new Repository(stack, 'Repo');
  new CfnOutput(stack, 'NpmRepositoryEndpoint', { value: repo.repositoryNpmEndpoint });

  // THEN
  expect(app.synth().getStackByName(stack.stackName).template).toMatchSnapshot({
    Parameters: expect.anything(),
    Resources: {
      AWS679f53fac002430cb0da5b7982bd22872D164C4C: {
        Properties: expect.anything(),
        Type: 'AWS::Lambda::Function',
      },
      AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2: {
        Properties: expect.anything(),
        Type: 'AWS::IAM::Role',
      },
      Repo: expect.anything(),
      RepoDomainC79FB030: expect.anything(),
    },
  });
});

test('S3 bucket ARN', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  // WHEN
  const repo = new Repository(stack, 'Repo');
  new CfnOutput(stack, 'S3BucketArn', { value: repo.s3BucketArn });

  // THEN
  expect(app.synth().getStackByName(stack.stackName).template).toMatchSnapshot({
    Parameters: expect.anything(),
    Resources: {
      AWS679f53fac002430cb0da5b7982bd22872D164C4C: {
        Properties: expect.anything(),
        Type: 'AWS::Lambda::Function',
      },
      AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2: {
        Properties: expect.anything(),
        Type: 'AWS::IAM::Role',
      },
      Repo: expect.anything(),
      RepoDomainC79FB030: expect.anything(),
    },
  });
});

test('grantReadFromRepository', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'TestStack');
  const role = new Role(stack, 'Role', { assumedBy: new ServicePrincipal('test.service-principal') });
  const repo = new Repository(stack, 'Repo');

  // WHEN
  repo.grantReadFromRepository(role);

  // THEN
  expect(stack).toHaveResource('AWS::IAM::Policy', {
    PolicyDocument: {
      Statement: [{
        Action: 'sts:GetServiceBearerToken',
        Condition: {
          StringEquals: { 'sts:AWSServiceName': 'codeartifact.amazonaws.com' },
        },
        Effect: 'Allow',
        Resource: '*',
      }, {
        Action: ['codeartifact:GetAuthorizationToken', 'codeartifact:GetRepositoryEndpoint', 'codeartifact:ReadFromRepository'],
        Effect: 'Allow',
        Resource: [
          stack.resolve(repo.repositoryDomainArn),
          stack.resolve(repo.repositoryArn),
        ],
      }],
      Version: '2012-10-17',
    },
    PolicyName: 'RoleDefaultPolicy5FFB7DAB',
    Roles: [stack.resolve(role.roleName)],
  });
});

test('throughVpcEndpoint', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'TestStack');
  const vpc = new Vpc(stack, 'VPC');
  const api = vpc.addInterfaceEndpoint('API', {
    service: new InterfaceVpcEndpointAwsService('codeartifact.api'),
  });
  const repositories = vpc.addInterfaceEndpoint('Repositories', {
    service: new InterfaceVpcEndpointAwsService('codeartifact.repositories'),
  });

  const role = new Role(stack, 'Role', { assumedBy: new ServicePrincipal('test.service-principal') });
  const repo = new Repository(stack, 'Repo');
  const vpcRepo = repo.throughVpcEndpoint(api, repositories);

  // WHEN
  vpcRepo.grantReadFromRepository(role);

  // THEN
  expect(vpcRepo.repositoryDomainOwner).toBe(repo.repositoryDomainOwner); // Example pass-through...

  expect(stack).toHaveResource('AWS::IAM::Policy', {
    PolicyDocument: {
      Statement: [{
        Action: 'sts:GetServiceBearerToken',
        Condition: {
          StringEquals: { 'sts:AWSServiceName': 'codeartifact.amazonaws.com' },
        },
        Effect: 'Allow',
        Resource: '*',
      }, {
        Action: ['codeartifact:GetAuthorizationToken', 'codeartifact:GetRepositoryEndpoint', 'codeartifact:ReadFromRepository'],
        Effect: 'Allow',
        Resource: [
          stack.resolve(repo.repositoryDomainArn),
          stack.resolve(repo.repositoryArn),
        ],
      }],
      Version: '2012-10-17',
    },
    PolicyName: 'RoleDefaultPolicy5FFB7DAB',
    Roles: [stack.resolve(role.roleName)],
  });

  expect(stack).toHaveResource('AWS::EC2::VPCEndpoint', {
    PolicyDocument: {
      Statement: [{
        Action: 'sts:GetServiceBearerToken',
        Condition: {
          StringEquals: { 'sts:AWSServiceName': 'codeartifact.amazonaws.com' },
        },
        Effect: 'Allow',
        Principal: { AWS: stack.resolve(role.roleArn) },
        Resource: '*',
      }, {
        Action: ['codeartifact:GetAuthorizationToken', 'codeartifact:GetRepositoryEndpoint'],
        Effect: 'Allow',
        Principal: { AWS: stack.resolve(role.roleArn) },
        Resource: [
          stack.resolve(repo.repositoryDomainArn),
          stack.resolve(repo.repositoryArn),
        ],
      }],
      Version: '2012-10-17',
    },
    ServiceName: stack.resolve(`com.amazonaws.${stack.region}.codeartifact.api`),
  });

  expect(stack).toHaveResource('AWS::EC2::VPCEndpoint', {
    PolicyDocument: {
      Statement: [{
        Action: 'codeartifact:ReadFromRepository',
        Effect: 'Allow',
        Principal: { AWS: stack.resolve(role.roleArn) },
        Resource: stack.resolve(repo.repositoryArn),
      }],
      Version: '2012-10-17',
    },
    ServiceName: stack.resolve(`com.amazonaws.${stack.region}.codeartifact.repositories`),
  });
});
