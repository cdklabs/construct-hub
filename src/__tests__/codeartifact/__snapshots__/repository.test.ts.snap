// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S3 bucket ARN 1`] = `
{
  "Mappings": {
    "DefaultCrNodeVersionMap": {
      "af-south-1": {
        "value": "nodejs16.x",
      },
      "ap-east-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-2": {
        "value": "nodejs16.x",
      },
      "ap-northeast-3": {
        "value": "nodejs16.x",
      },
      "ap-south-1": {
        "value": "nodejs16.x",
      },
      "ap-south-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-1": {
        "value": "nodejs16.x",
      },
      "ap-southeast-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-3": {
        "value": "nodejs16.x",
      },
      "ca-central-1": {
        "value": "nodejs16.x",
      },
      "cn-north-1": {
        "value": "nodejs16.x",
      },
      "cn-northwest-1": {
        "value": "nodejs16.x",
      },
      "eu-central-1": {
        "value": "nodejs16.x",
      },
      "eu-central-2": {
        "value": "nodejs16.x",
      },
      "eu-north-1": {
        "value": "nodejs16.x",
      },
      "eu-south-1": {
        "value": "nodejs16.x",
      },
      "eu-south-2": {
        "value": "nodejs16.x",
      },
      "eu-west-1": {
        "value": "nodejs16.x",
      },
      "eu-west-2": {
        "value": "nodejs16.x",
      },
      "eu-west-3": {
        "value": "nodejs16.x",
      },
      "me-central-1": {
        "value": "nodejs16.x",
      },
      "me-south-1": {
        "value": "nodejs16.x",
      },
      "sa-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-2": {
        "value": "nodejs16.x",
      },
      "us-gov-east-1": {
        "value": "nodejs16.x",
      },
      "us-gov-west-1": {
        "value": "nodejs16.x",
      },
      "us-iso-east-1": {
        "value": "nodejs14.x",
      },
      "us-iso-west-1": {
        "value": "nodejs14.x",
      },
      "us-isob-east-1": {
        "value": "nodejs14.x",
      },
      "us-west-1": {
        "value": "nodejs16.x",
      },
      "us-west-2": {
        "value": "nodejs16.x",
      },
    },
  },
  "Outputs": {
    "S3BucketArn": {
      "Value": {
        "Fn::GetAtt": [
          "RepoDescribeDomainBFB922C1",
          "domain.s3BucketArn",
        ],
      },
    },
  },
  "Parameters": Anything,
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Anything,
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": Anything,
      "Type": "AWS::IAM::Role",
    },
    "Repo": Anything,
    "RepoDescribeDomainBFB922C1": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "RepoDescribeDomainCustomResourcePolicy66761D14",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"CodeArtifact","action":"describeDomain","parameters":{"domain":"",
              {
                "Fn::GetAtt": [
                  "Repo",
                  "DomainName",
                ],
              },
              "","domainOwner":"",
              {
                "Fn::GetAtt": [
                  "Repo",
                  "DomainOwner",
                ],
              },
              ""},"physicalResourceId":{"responsePath":"domain.s3BucketArn"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::CoreArtifactDomainDescription",
      "UpdateReplacePolicy": "Delete",
    },
    "RepoDescribeDomainCustomResourcePolicy66761D14": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codeartifact:DescribeDomain",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "RepoDomainC79FB030",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RepoDescribeDomainCustomResourcePolicy66761D14",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RepoDomainC79FB030": Anything,
    "RepoPublishing9FE3D602": {
      "Properties": {
        "Description": "Publishing repository",
        "DomainName": {
          "Fn::GetAtt": [
            "RepoDomainC79FB030",
            "Name",
          ],
        },
        "RepositoryName": "c824e6f26ac4b798fe181c4568a8a55d69111ad893-publish-overlay",
      },
      "Type": "AWS::CodeArtifact::Repository",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`npm repository endpoint 1`] = `
{
  "Mappings": {
    "DefaultCrNodeVersionMap": {
      "af-south-1": {
        "value": "nodejs16.x",
      },
      "ap-east-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-2": {
        "value": "nodejs16.x",
      },
      "ap-northeast-3": {
        "value": "nodejs16.x",
      },
      "ap-south-1": {
        "value": "nodejs16.x",
      },
      "ap-south-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-1": {
        "value": "nodejs16.x",
      },
      "ap-southeast-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-3": {
        "value": "nodejs16.x",
      },
      "ca-central-1": {
        "value": "nodejs16.x",
      },
      "cn-north-1": {
        "value": "nodejs16.x",
      },
      "cn-northwest-1": {
        "value": "nodejs16.x",
      },
      "eu-central-1": {
        "value": "nodejs16.x",
      },
      "eu-central-2": {
        "value": "nodejs16.x",
      },
      "eu-north-1": {
        "value": "nodejs16.x",
      },
      "eu-south-1": {
        "value": "nodejs16.x",
      },
      "eu-south-2": {
        "value": "nodejs16.x",
      },
      "eu-west-1": {
        "value": "nodejs16.x",
      },
      "eu-west-2": {
        "value": "nodejs16.x",
      },
      "eu-west-3": {
        "value": "nodejs16.x",
      },
      "me-central-1": {
        "value": "nodejs16.x",
      },
      "me-south-1": {
        "value": "nodejs16.x",
      },
      "sa-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-2": {
        "value": "nodejs16.x",
      },
      "us-gov-east-1": {
        "value": "nodejs16.x",
      },
      "us-gov-west-1": {
        "value": "nodejs16.x",
      },
      "us-iso-east-1": {
        "value": "nodejs14.x",
      },
      "us-iso-west-1": {
        "value": "nodejs14.x",
      },
      "us-isob-east-1": {
        "value": "nodejs14.x",
      },
      "us-west-1": {
        "value": "nodejs16.x",
      },
      "us-west-2": {
        "value": "nodejs16.x",
      },
    },
  },
  "Outputs": {
    "NpmRepositoryEndpoint": {
      "Value": {
        "Fn::GetAtt": [
          "RepoGetEndpoint5034567A",
          "repositoryEndpoint",
        ],
      },
    },
  },
  "Parameters": Anything,
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Anything,
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": Anything,
      "Type": "AWS::IAM::Role",
    },
    "Repo": Anything,
    "RepoDomainC79FB030": Anything,
    "RepoGetEndpoint5034567A": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "RepoGetEndpointCustomResourcePolicy40405D0D",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"CodeArtifact","action":"getRepositoryEndpoint","parameters":{"domain":"",
              {
                "Fn::GetAtt": [
                  "Repo",
                  "DomainName",
                ],
              },
              "","domainOwner":"",
              {
                "Fn::GetAtt": [
                  "Repo",
                  "DomainOwner",
                ],
              },
              "","format":"npm","repository":"",
              {
                "Fn::GetAtt": [
                  "Repo",
                  "Name",
                ],
              },
              ""},"physicalResourceId":{"responsePath":"repositoryEndpoint"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": {
          "Fn::Join": [
            "",
            [
              "{"service":"CodeArtifact","action":"getRepositoryEndpoint","parameters":{"domain":"",
              {
                "Fn::GetAtt": [
                  "Repo",
                  "DomainName",
                ],
              },
              "","domainOwner":"",
              {
                "Fn::GetAtt": [
                  "Repo",
                  "DomainOwner",
                ],
              },
              "","format":"npm","repository":"",
              {
                "Fn::GetAtt": [
                  "Repo",
                  "Name",
                ],
              },
              ""},"physicalResourceId":{"responsePath":"repositoryEndpoint"}}",
            ],
          ],
        },
      },
      "Type": "Custom::CodeArtifactNpmRepositoryEndpoint",
      "UpdateReplacePolicy": "Delete",
    },
    "RepoGetEndpointCustomResourcePolicy40405D0D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codeartifact:GetRepositoryEndpoint",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "Repo",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RepoGetEndpointCustomResourcePolicy40405D0D",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RepoPublishing9FE3D602": {
      "Properties": {
        "Description": "Publishing repository",
        "DomainName": {
          "Fn::GetAtt": [
            "RepoDomainC79FB030",
            "Name",
          ],
        },
        "RepositoryName": "c824e6f26ac4b798fe181c4568a8a55d69111ad893-publish-overlay",
      },
      "Type": "AWS::CodeArtifact::Repository",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
